tests/parser/run-parser-tests.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+// Throw exception if anything fails.
+set_error_handler(
+	function ( $severity, $message, $file, $line ) {
+		throw new ErrorException( $message, 0, $severity, $file, $line );
+	}
+);
+
+require_once __DIR__ . '/../../wp-includes/mysql/class-mysql-lexer.php';
+require_once __DIR__ . '/../../wp-includes/parser/class-parser.php';
+
+function getStats( $total, $failures, $exceptions ) {
+	return sprintf(
+		'Total: %5d  |  Failures: %4d / %2d%%  |  Exceptions: %4d / %2d%%',
+		$total,
+		$failures,
+		$failures / $total * 100,
+		$exceptions,
+		$exceptions / $total * 100
+	);
+}
+
+$grammar_data = include __DIR__ . '/../../wp-includes/mysql/mysql-grammar.php';
+$grammar      = new Grammar( $grammar_data );
+
+$handle     = fopen( __DIR__ . '/data/queries.csv', 'r' );
+$i          = 1;
+$failures   = array();
+$exceptions = array();
+while ( ( $query = fgetcsv( $handle ) ) !== false ) {
+	$query = $query[0];
+	if ( null === $query ) {
+		continue;
+	}
+
+	// Skip overflow queries for now.
+	if (
+		str_contains( $query, 'func_overflow()' )
+		|| str_contains( $query, 'proc_overflow()' )
+		|| str_contains( $query, 'table_overflow()' )
+		|| str_contains( $query, 'trigger_overflow' )
+	) {
+		continue;
+	}
+
+	try {
+		$tokens = tokenize_query( $query );
+		if ( empty( $tokens ) ) {
+			throw new Exception( 'Empty tokens' );
+		}
+
+		$parser     = new Parser( $grammar, $tokens );
```

Comment:
How about `WP_MySQL_Parser::parse( $tokens, $grammar )`? It's compliant with how WordPress names its classes, which makes it easier to propose for core later on. Also, the `$grammar` argument could be optional and the static method could know how to load the default one, e.g. by calling `WP_MySQL_Grammar::for_version( "8.0" )` or so. I know static calls are frowned upon in TDD, but I don't think we'll ever want to mock either of these classes.

Reply:
@adamziel 👍 for `WP_MySQL_Lexer` and `WP_MySQL_Parser`, but I'm wondering about the fact that the parser is universal (apart from the hardcoded lookeahead).

Maybe, it could be something like `class WP_MySQL_Parser extends WP_Parser`, what do you think? And maybe it's not necessary to do this now, we can always split it in the future. But then, what about `ParseTree` and `Grammar` classes? They are not MySQL-specific either, so in the end, maybe it makes sense to do `class WP_MySQL_Parser extends WP_Parser` now? The structure would then be:

```
parser/class-wp-parser.php
parser/class-wp-parse-tree.php
parser/class-wp-parser-grammar.php
mysql/class-wp-mysql-parser.php (extends class-wp-parser.php)
mysql/class-wp-mysql-lexer.php
mysql/class-wp-mysql-token.php
mysql/mysql-grammar.php (or wp-mysql-grammar?)
```

Reply:
Good point! And sounds good